engine/schema/src/main/java/com/cloud/vm/NicVO.java

```diff
@@ -394,4 +397,14 @@ public String getNsxLogicalSwitchPortUuid() {
     public void setNsxLogicalSwitchPortUuid(String nsxLogicalSwitchPortUuid) {
         this.nsxLogicalSwitchPortUuid = nsxLogicalSwitchPortUuid;
     }
+
```

Comment:
Sorry, I don't understand. Can you please explain?

Reply:
@SadiJr 
can you point out where two NicVO are compared in the source code ? is it possible to compare using `id` ?

one more question, if the nic is removed in cloudstack, then restore VM from veeam, will the nic be marked as not removed ?

Reply:
@weizhouapache 

I'm sorry for the late reply. On line 909 of the VMWareGuru class, the remove method of a list object is called, passing a NicVO object as a reference. This is done to remove the NIC in question from the list of NICs to be removed from the VM (line 919); that is, NICs that are not in the list will be kept in the VM. It turns out that the NicVO object doesn't have the equals and hashcode methods overridden. Thus, the native method of Java is used, which uses the object reference to check if they are the same. With this, even if the NICs found in the database are the same discovered in VMware after a restore process, ACS will think they are different due to being two different object instances (objects of the same class), and will remove them of the VM.

About removed NICs, actually the NIC was not marked as not removed. I fixed that and thank you for pointing out this error.

PS: I don't know if I was able to express myself correctly (communication is not exactly one of my strong points), let me know if you don't understand something.

Reply:
@SadiJr 
my suggestion is, we could replace  `allNics` with `allNicIds` which contains the id of the vm nics, 

the other lines can be changed to

```
            if (nicVO != null) {
                allNicIds.remove(nicVO.getId());
            }
```
and
```
        for (final Long unMappedNicId : allNicIds) {
            vmManager.removeNicFromVm(vm, _nicDao.findById(unMappedNicId));
        }
```

Reply:
@weizhouapache yes, I can do that, but I think is better to override the equals and hashCode in NicVO, to make sure this error don't appears in other parts of code. What you think about?

Reply:
@SadiJr I get what you are saying and I'm up for both the solutions as you and @weizhouapache suggested.
My only concern with the equals method override is if it causes any regression in other places. To avoid such scenarios @weizhouapache 's solution makes it safer. 

@weizhouapache any suggestions here!

Reply:
@harikrishna-patnala I don't see how this change could break other parts of the code, if two nicVO objects were being compared and they came out as equal before, they'll still continue doing so.